### DISCUSSION TITLE
Cascade delete not working depending on how deletion is made

### DISCUSSION THREAD

**Discussion by emontarsoloamf** [sqlalchemy/sqlalchemy#7974] [Usage Questions] [answered]
python version : 3.7.10
sql alchemy version : 1.4.11

Hi,
I'm currently working on a project where the database purging is very slow. Currently, the code executes a query to get the objects to delete , and then iterate on each object to apply .delete() method. 
```
to_del = session.query(MyBase).filter(MyBase.data < date_purge).all()
for x in to_del:
    session.delete(x)
    session.commit()

```

I tried to change this part, applying directly the delete method to the query :

`session.query(MyBase).filter(MyBase.date < date_purge).delete()`

But this raises an Integrity error :

 `IntegrityError: ('23000', '[23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The DELETE statement conflicted with the REFERENCE constraint`

Even though the bases are declared with "cascade delete" statements, and the previous method for the **same instance** works perfectly : with the first method instances from MyBase are deleted, and so are their children instances.

I tried to reproduce the error, with a more simple, easy-to-reproduce base;  but couldnt find a way to have the exact same error.  Though, I noticed cascade deletion does not work the same depending on how you delete your instances.

Here is a simple code for creating a database with parent/children relationships:
```py
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, sessionmaker
from sqlalchemy.schema import CreateSchema, DropSchema, MetaData
from sqlalchemy import Column, ForeignKey, Integer, String, create_engine, UniqueConstraint
from sqlalchemy import delete


Base = declarative_base()

class Painting(Base):
    __tablename__ = "painting"
    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(450))
    painter_id = Column(Integer, ForeignKey("painter.id", ondelete='CASCADE'))
    __table_args__ = (UniqueConstraint("title"),)
    def __repr__(self):
        return (f"'{self.title}'(by {self.painter.lastname})")


class Painter(Base):
    __tablename__ = "painter"
    id = Column(Integer, primary_key=True, autoincrement=True)
    firstname = Column(String(450))
    lastname = Column(String(450))
    paintings = relationship("Painting", cascade="all,delete", backref="painter")
    movement_id = Column(Integer, ForeignKey("movement.id", ondelete='CASCADE'))
    def __repr__(self):
        return (f"{self.firstname} {self.lastname} ({self.movement.name})")
    
class Movement(Base):
    __tablename__ = "movement"
    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(450))
    painters = relationship("Painter", cascade="all,delete", backref="movement")
    __table_args__ = (UniqueConstraint("name"),)
    def __repr__(self):
        return (f"{self.name} ({', '.join([p.lastname for p in self.painters])})")

engine = create_engine("sqlite:///test.db")
engine.connect()
Base.metadata.create_all(engine)

movements = ["Impressionisme", "Fauvisme", "Réalisme"]
painters = [(("Henri", "Matisse"), "Fauvisme"),
           (("Claude", "Monet"), "Impressionisme"),
           (("Berthe", "Morisot"), "Impressionisme"),
           (("Gustave", "Courbet"), "Réalisme")]
paintings = [("La femme au chapeau", ("Henri", "Matisse")),
             ("La blouse Romaine", ("Henri", "Matisse")),
             ("Les nymphéas", ("Claude", "Monet")),
             ("Le berceau", ("Berthe", "Morisot")),
             ("Autoportrait", ("Gustave", "Courbet"))]

Session = sessionmaker(engine)
id_mvmts, id_painters = {}, {}
with Session() as session:
    for name in movements :
        mvmt = Movement(name=name)
        session.add(mvmt)
        session.flush()
        id_mvmts[name]= mvmt.id
    
    for names, mvmt in painters:
        painter = Painter(firstname=names[0], lastname=names[1], movement_id = id_mvmts[mvmt])
        session.add(painter)
        session.flush()
        id_painters[" ".join(names)]= painter.id
        
    for title, names in paintings :
        painting = Painting(title=title, painter_id=id_painters[" ".join(names)])
        session.add(painting)
        session.flush()

    session.commit()
    session.close()
```
In the end the data are the following : 
```
print(session.query(Movement).all())
>>>[Impressionisme (Monet, Morisot), Fauvisme (Matisse), Réalisme (Courbet)]
print(session.query(Painter).all())
>>>[Henri Matisse (Fauvisme), Claude Monet (Impressionisme), Berthe Morisot (Impressionisme), Gustave Courbet (Réalisme)]
print(session.query(Painting).all())
>>>['La femme au chapeau'(by Matisse), 'La blouse Romaine'(by Matisse), 'Les nymphéas'(by Monet), 'Le berceau'(by Morisot), 'Autoportrait'(by Courbet)]

```

Let's say I want to delete one Movement instance , the expected result will be : Movement instance is deleted, and so are the related Painters and Paintings. This result is achieved with the first method:
```
with Session() as session:
    to_del = session.query(Movement).filter(Movement.name == "Impressionisme").all()
    for m in to_del:
        session.delete(m)
        session.commit()

print(session.query(Movement).all())
print(session.query(Painter).all())
print(session.query(Painting).all())
>>>[Fauvisme (Matisse), Réalisme (Courbet)]
>>>[Henri Matisse (Fauvisme), Gustave Courbet (Réalisme)]
>>>['La femme au chapeau'(by Matisse), 'La blouse Romaine'(by Matisse), 'Autoportrait'(by Courbet)]
```
But if we apply the .delete() directly to the query, only Movement Instance is deleted, and Painter.movement is set to None (leading to attribute errors...).
```
with Session() as session:
    session.query(Movement).filter(Movement.name == "Impressionisme").delete()
    session.commit()

```
Same issue with this code:
```
with Session() as session:
    d = delete(Movement).filter(Movement.name == "Impressionisme")
    engine.execute(d)

```

What I am doing wrong here ?
Thanks,

**Comment by zzzeek** [comment, marked as answer]
So the query.delete() method emits a SQL DELETE statement directly to the database when you call it.   When this happens, the SQLAlchemy ORMs [unit of work](https://docs.sqlalchemy.org/en/14/glossary.html?highlight=unit%20work#term-unit-of-work) process is not involved in any way; this is normally the process that goes through the often non-performant task of loading all the related objects for a particular object upon which you called delete(), organizing them up and then emitting "DELETE" statements for all those related objects individually, and it is indeed very inefficient.

so when you run SQL DELETE directly, you instead instruct your database to delete all the related rows directly.    from SQLAlchemy you configure [ON DELETE CASCADE](https://docs.sqlalchemy.org/en/14/core/constraints.html?highlight=delete%20cascade#on-update-and-on-delete) for your foreign key constraints and ensure the schema in your SQL Server database has these constraints also (if this is an existing production database, you'd need to apply database migrations to add these directives).  then the SQL DELETE statement can run on the primary objects and SQL Server takes care of the rest.


guidelines for how to configure your relationships to work smoothly with ON DELETE CASCADE is at https://docs.sqlalchemy.org/en/14/orm/cascades.html#passive-deletes .

the general info about query.delete() limitations, including this one, are in the big pink dragon box at the bottom of the section https://docs.sqlalchemy.org/en/14/orm/session_basics.html#update-and-delete-with-arbitrary-where-clause

**Reply by zzzeek** [reply]
looks like you have CASCADE set up for some of those foreign key constraints so I would ensure that these directives exist in the database correctly as corresponds to the particular DELETE statement that's causing a problem.

**Reply by emontarsoloamf** [reply]
Thanks for your reply.
My database is currently created through sqlalchemy, with commands : 
```
engine = create_engine("sqlite:///test.db") 
engine.connect()
Base.metadata.create_all(engine)
```
Isn't the parameter `cascade="all,delete"` within the relationship function enough to ensure a Delete Cascade rule ? (I don't know how to check if this constraint exist for SQL, but I know my tables were created with this parameter).
Do I need to create my tables with SQL scripts first if I want the cascade delete rules to work ?

Edit : I tried to delete one row with a SQL script (applied from SQL manager) and got the same error, which confirms that the cascade delete rule does not exist for SQL alone.

**Reply by zzzeek** [reply]
SQLite does not use foreign key constraints unless you enable them on a per connection basis.  follow the recipe at https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#foreign-key-support
